src/main/scala/in/ashwanthkumar/suuchi/rpc/Router.scala

```diff
@@ -0,0 +1,58 @@
+package in.ashwanthkumar.suuchi.rpc
+
+import io.grpc.ServerCall.Listener
+import io.grpc._
+import io.grpc.netty.NettyChannelBuilder
+import io.grpc.stub.ClientCalls
+import org.slf4j.LoggerFactory
+
+/**
+ * Router decides to route the incoming request to right node in the cluster as defined
+ * by the [[RoutingStrategy]].
+ *
+ * @param routingStrategy
+ */
+class Router(routingStrategy: RoutingStrategy) extends ServerInterceptor {
+  private val log = LoggerFactory.getLogger(getClass)
+
+  override def interceptCall[ReqT, RespT](serverCall: ServerCall[ReqT, RespT], headers: Metadata, next: ServerCallHandler[ReqT, RespT]): Listener[ReqT] = {
+    log.debug("Intercepting " + serverCall.getMethodDescriptor.getFullMethodName + " method")
+    new Listener[ReqT] {
+      val delegate = next.startCall(serverCall, headers)
+      var forwarded = false
+
+      override def onReady(): Unit = delegate.onReady()
+      override def onMessage(incomingRequest: ReqT): Unit = {
+        // TODO - Handle forwarding loop here
+        if(routingStrategy.route.isDefinedAt(incomingRequest)) {
```

Comment:
Alright. We should then use `applyOrElse` as that seems a better way to do this (from the scaladoc)

> Note that expression `pf.applyOrElse(x, default)` is equivalent to {{{ if(pf isDefinedAt x) pf(x) else default(x) }}} except that `applyOrElse` method can be implemented more efficiently.